runtime/lua/vim/uri.lua

```diff
@@ -10,62 +14,59 @@ local tohex = require('bit').tohex
 local URI_SCHEME_PATTERN = '^([a-zA-Z]+[a-zA-Z0-9.+-]*):.*'
 local WINDOWS_URI_SCHEME_PATTERN = '^([a-zA-Z]+[a-zA-Z0-9.+-]*):[a-zA-Z]:.*'
 local PATTERNS = {
-  ---RFC 2396
-  ---https://tools.ietf.org/html/rfc2396#section-2.2
+  --- RFC 2396 https://tools.ietf.org/html/rfc2396#section-2.2
   rfc2396 = "^A-Za-z0-9%-_.!~*'()",
-  ---RFC 2732
-  ---https://tools.ietf.org/html/rfc2732
+  --- RFC 2732 https://tools.ietf.org/html/rfc2732
   rfc2732 = "^A-Za-z0-9%-_.!~*'()[]",
-  ---RFC 3986
-  ---https://tools.ietf.org/html/rfc3986#section-2.2
+  --- RFC 3986 https://tools.ietf.org/html/rfc3986#section-2.2
   rfc3986 = "^A-Za-z0-9%-._~!$&'()*+,;=:@/",
 }
 
----Converts hex to char
----@param hex string
----@return string
+--- Converts hex to char
+--- @param hex string
+--- @return string
 local function hex_to_char(hex)
   return schar(tonumber(hex, 16))
 end
 
----@param char string
----@return string
+--- @param char string
+--- @return string
 local function percent_encode_char(char)
   return '%' .. tohex(sbyte(char), 2)
 end
 
----@param uri string
----@return boolean
+--- @param uri string
+--- @return boolean
 local function is_windows_file_uri(uri)
   return uri:match('^file:/+[a-zA-Z]:') ~= nil
 end
 
----URI-encodes a string using percent escapes.
----@param str string string to encode
----@param rfc "rfc2396" | "rfc2732" | "rfc3986" | nil
----@return string encoded string
-function M.uri_encode(str, rfc)
+--- URI-encodes a string.
+--- @param str string string to encode
+--- @param rfc "rfc2396" | "rfc2732" | "rfc3986" | nil
+--- @return string encoded string
+function M.encode(str, rfc)
   local pattern = PATTERNS[rfc] or PATTERNS.rfc3986
   return (str:gsub('([' .. pattern .. '])', percent_encode_char)) -- clamped to 1 retval with ()
 end
 
----URI-decodes a string containing percent escapes.
----@param str string string to decode
----@return string decoded string
-function M.uri_decode(str)
+--- Decodes a URI-encoded string.
+--- @param str string string to decode
+--- @return string decoded string
+function M.decode(str)
   return (str:gsub('%%([a-fA-F0-9][a-fA-F0-9])', hex_to_char)) -- clamped to 1 retval with ()
 end
 
----Gets a URI from a file path.
----@param path string Path to file
----@return string URI
-function M.uri_from_fname(path)
+--- Gets a URI from a file path.
+--- @param path string Path to file
+--- @return string URI
+function M.from_fname(path)
   local volume_path, fname = path:match('^([a-zA-Z]:)(.*)')
   local is_windows = volume_path ~= nil
   if is_windows then
-    path = volume_path .. M.uri_encode(fname:gsub('\\', '/'))
+    path = volume_path .. M.encode(fname:gsub('\\', '/'))
```

Comment:
would it be possible to include correct uri encoding of the volume_path and possibly add a vscode-like normalization?
```
volume_path = volume_path:lower():gsub(":", "%%3A") 
```

Reply:
Also I think that the M.encode should use pattern `"^A-Za-z0-9%-_.!~*'()/"` (rfc2396 + "/" = keep "unreserved chars" and slashes), so that odd filenames like `#@file@#` are not translated into an uri-fragment.